File: PTFMXMLParser.m
//
//  PTFMXMLParser.m
//  XRose
//
//  Created by Tom Moore on 12/28/05.
//
// MIT License
//
// Copyright (c) 2005 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "PTFMXMLParser.h"
#import "sqlite3.h"

@implementation PTFMXMLParser

-(id)initWithXMLAtPath:(NSString *)path
{
	if (!(self = [super init])) return nil;
	if(self)
	{
		NSURL *theURL = [NSURL fileURLWithPath:path];
		tableDefinition = [[NSMutableArray alloc] init];
		dataRows = [[NSMutableArray alloc] init];
		correctType = NO;
		theParser = [[NSXMLParser alloc] initWithContentsOfURL:theURL];
		[theParser setDelegate:self];
		isValid = [theParser parse];
		tableStep = 0;
		
	}
	return self;
}


- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary *)attributeDict 
{

	if([elementName isEqualToString:@"FMPXMLRESULT"])
		correctType = YES;
	else if([elementName isEqualToString:@"DATABASE"])
	{
		tableName = [attributeDict objectForKey:@"NAME"];
		//NSLog(@"Table Name: %@",tableName);
	}
	else if([elementName isEqualToString:@"FIELD"])
	{
		[tableDefinition addObject:attributeDict];
	}
	else if([elementName isEqualToString:@"ROW"])
	{
		currentRow = [[NSMutableArray alloc] init];
	}
	else if([elementName isEqualToString:@"COL"])
	{
		hasData = NO;
	}
	else if([elementName isEqualToString:@"DATA"])
	{
		hasData = YES;
		currentContents = [[NSMutableString alloc] init];
	}
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName 
{
	if([elementName isEqualToString:@"ROW"])
	{
		[dataRows addObject:[NSArray arrayWithArray:currentRow]];
		currentRow = nil;
	}
	else if([elementName isEqualToString:@"COL"])
	{
		if(!hasData)
			[currentRow addObject:[NSNull null]];
	}
	else if([elementName isEqualToString:@"DATA"])
	{
		[currentRow addObject:currentContents];
		currentContents = nil;
	}
}

- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string
{

	if(currentContents)
		[currentContents appendString:string];
}

-(BOOL)isCorrectType
{
	return correctType;
}

-(BOOL)isValid
{
	return isValid;
}

-(NSString *)sqliteTableSchema
{
	NSMutableString *theString = [[NSMutableString alloc] init];
	[theString appendFormat:@"CREATE TABLE \"%@\" (\n_id INTEGER PRIMARY KEY,\n",tableName];
	NSDictionary *aDict;
	for(int i=0;i<[tableDefinition count];i++)
	{
		aDict = [tableDefinition objectAtIndex:i];
		if(i == ([tableDefinition count] -1 )) 
			[theString appendFormat:@"%@ %@\n )",[aDict objectForKey:@"NAME"], [aDict objectForKey:@"TYPE"]];
		else
			[theString appendFormat:@"%@ %@, \n",[aDict objectForKey:@"NAME"], [aDict objectForKey:@"TYPE"]];
	}
	return theString;
}

-(void)stepTableName
{
	//This is designed to step the table name if table name already exists
	if(tableStep == 0)
	{
		NSString *oldName = tableName;
		tableName = [oldName stringByAppendingString:@"_1"];
		oldName = nil;
		tableStep++;
	}
	else
	{
		NSString *oldName = tableName;
		NSRange aRange,aIntersectRange,aStringRange;
		aStringRange.length = [oldName length];
		aStringRange.location = 0;
		aRange = [oldName rangeOfString:[NSString stringWithFormat:@"_%i",tableStep] options:NSBackwardsSearch];
		aIntersectRange = NSIntersectionRange(aStringRange,aRange);
		tableStep++;
		tableName = [[oldName substringWithRange:aIntersectRange] stringByAppendingFormat:@"_%i",tableStep];
		oldName = nil;
		tableStep++;
	}
	return;
}

-(int)rows
{
	return (int)[dataRows count];
}

-(NSString *)sqliteCommandForRow:(int)aRow
{
	NSMutableString *command = [[NSMutableString alloc] init];
	NSArray *rowArray = [dataRows objectAtIndex:aRow];
	[command appendString:[self assembleSQLITEInsertBase]];
	int count;
	count = (int)[rowArray count];
	[command appendString:@"VALUES ("];
	for(int i = 0;i<count;i++)
	{
		if([[rowArray objectAtIndex:i] isKindOfClass:[NSNull class]])
			[command appendString:@"NULL"];
		else
			[command appendFormat:@"\"%@\"",[rowArray objectAtIndex:i]];
		if(i<(count-1))
			[command appendString:@","];
		else
			[command appendString:@")"];
	}
	return command;
}

-(NSString *)assembleSQLITEInsertBase
{
	NSMutableString *base = [[NSMutableString alloc] init];
	[base appendFormat:@"INSERT INTO \"%@\" (",tableName];

	for(int i=0;i<[tableDefinition count];i++)
	{
		if(i==([tableDefinition count] - 1))
			[base appendFormat:@"%@) ",[[tableDefinition objectAtIndex:i] objectForKey:@"NAME"]];
		else
			[base appendFormat:@"%@,",[[tableDefinition objectAtIndex:i] objectForKey:@"NAME"]];
	}
	return base;
}

-(BOOL)writeToSQLITE:(sqlite3 *)db withError:(NSError **)anError
{
	int error;
	int count;
	sqlite3_stmt *stmt;
	const char *pzTail;
	char *errorMsg;

	//make sure table name is unique
	count = -1;//set it up for failure on a while statement
	
	while(count!=0)
	{
		NSString *aString = [NSString stringWithFormat:@"select count(*) from sqlite_master where tbl_name=\"%@\"",tableName];
		sqlite3_prepare(db,[aString UTF8String],-1,&stmt,&pzTail);
		while(sqlite3_step(stmt)==SQLITE_ROW)
		{
			count = sqlite3_column_int(stmt,0);
		}
		sqlite3_finalize(stmt);
		if(count>0)
			[self stepTableName];
	}
	
	
	error = sqlite3_exec(db,"BEGIN",NULL,NULL,&errorMsg);
	if(error!=SQLITE_OK)
	{
        if(anError != NULL ) {
		*anError = [NSError errorWithDomain:@"SQLITE" code:error userInfo:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:[NSString stringWithUTF8String:(char *)sqlite3_errmsg(db)],nil]
																								forKeys:[NSArray arrayWithObjects:NSLocalizedDescriptionKey,nil]]];
                              }
		sqlite3_exec(db,"ROLLBACK",NULL,NULL,&errorMsg);
		return NO;
	}
	//now create table schema in db
	
	error = sqlite3_exec(db,[[self sqliteTableSchema] UTF8String],NULL,NULL,&errorMsg);
	//NSLog([self sqliteTableSchema]);
	if(error!=SQLITE_OK)
	{
        if(anError != NULL ){
            *anError = [NSError errorWithDomain:@"SQLITE" code:error userInfo:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:[NSString stringWithUTF8String:(char *)sqlite3_errmsg(db)],nil]
                                                                                                          forKeys:[NSArray arrayWithObjects:NSLocalizedDescriptionKey,nil]]];
        }
		
		sqlite3_exec(db,"ROLLBACK",NULL,NULL,&errorMsg);
		return NO;
	}
	
	//now insert rows
	for(int i=0;i<[self rows];i++)
	{
		error = sqlite3_exec(db,[[self sqliteCommandForRow:i] UTF8String],NULL,NULL,&errorMsg);
		if(error!=SQLITE_OK)
		{
            if(anError != NULL ) {
                *anError = [[NSError alloc] initWithDomain:@"SQLITE"
                                                      code:error
                                                  userInfo:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:[NSString stringWithUTF8String:(char *)sqlite3_errmsg(db)],nil]
                                                                                       forKeys:[NSArray arrayWithObjects:NSLocalizedDescriptionKey,nil]]];
            }
                sqlite3_exec(db,"ROLLBACK",NULL,NULL,&errorMsg);
			return NO;
		}
	}
	//commit changes
	
	error = sqlite3_exec(db,"COMMIT",NULL,NULL,&errorMsg);
	if(error!=SQLITE_OK)
	{
		if(anError != NULL ) {
		*anError = [NSError errorWithDomain:@"SQLITE" code:error userInfo:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:[NSString stringWithUTF8String:(char *)sqlite3_errmsg(db)],nil]
																									  forKeys:[NSArray arrayWithObjects:NSLocalizedDescriptionKey,nil]]];
        }
		sqlite3_exec(db,"ROLLBACK",NULL,NULL,&errorMsg);
		return NO;
	}
	
	return YES;
}
@end

File: PaleoRose/Classes/Table Importing/XML Parser/PTFMXMLParser.swift
// PTFMXMLParser.swift
// XRose
//
// Created by Tom Moore on 12/28/05.
//
// MIT License
//
// Copyright (c) 2005 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Cocoa
import SQLite3

@objc class PTFMXMLParser: NSObject, XMLParserDelegate {
    // MARK: - Properties
    
    private var tableDefinition: [[String: String]] = []
    private var dataRows: [[Any]] = []
    private var isValid: Bool = false
    private var correctType: Bool = false
    private var parser: XMLParser
    private var currentRow: [Any]?
    private var currentContents: NSMutableString?
    private var hasData: Bool = false
    private var tableName: String = ""
    private var tableStep: Int = 0
    
    // MARK: - Initialization
    
    @objc init?(xmlAtPath path: String) {
        guard let url = URL(fileURLWithPath: path) else { return nil }
        guard let xmlParser = XMLParser(contentsOf: url) else { return nil }
        
        self.parser = xmlParser
        super.init()
        
        parser.delegate = self
        isValid = parser.parse()
    }
    
    // MARK: - XMLParserDelegate
    
    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String] = [:]) {
        switch elementName {
        case "FMPXMLRESULT":
            correctType = true
            
        case "DATABASE":
            tableName = attributeDict["NAME"] ?? ""
            
        case "FIELD":
            tableDefinition.append(attributeDict)
            
        case "ROW":
            currentRow = []
            
        case "COL":
            hasData = false
            
        case "DATA":
            hasData = true
            currentContents = NSMutableString()
            
        default:
            break
        }
    }
    
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        switch elementName {
        case "ROW":
            if let row = currentRow {
                dataRows.append(row)
                currentRow = nil
            }
            
        case "COL":
            if !hasData {
                currentRow?.append(NSNull())
            }
            
        case "DATA":
            if let contents = currentContents as? String {
                currentRow?.append(contents)
            } else {
                currentContents = nil
            }

        default:
            break
        }
    }
    
    func parser(_ parser: XMLParser, foundCharacters string: String) {
        currentContents?.append(string)
    }
    
    // MARK: - Public API
    
    @objc var isCorrectType: Bool {
        correctType
    }
    
    @objc var isValidParse: Bool {
        isValid
    }
    
    @objc var rowCount: Int {
        dataRows.count
    }
    
    @objc func stepTableName() {
        if tableStep == 0 {
            tableName += "_1"
            tableStep += 1
        } else {
            if let range = tableName.range(of: "_\(tableStep)", options: .backwards) {
                tableName = String(tableName[..<range.lowerBound]) + "_\(tableStep + 1)"
            }
            tableStep += 1
        }
    }
    
    @objc func sqliteTableSchema() -> String {
        var schema = "CREATE TABLE \"\(tableName)\" (\n_id INTEGER PRIMARY KEY,\n"
        
        for (index, field) in tableDefinition.enumerated() {
            let name = field["NAME"] ?? ""
            let type = field["TYPE"] ?? ""
            
            if index == tableDefinition.count - 1 {
                schema += "\(name) \(type)\n)"
            } else {
                schema += "\(name) \(type),\n"
            }
        }
        
        return schema
    }
    
    @objc func sqliteCommand(forRow row: Int) -> String {
        guard row < dataRows.count else { return "" }
        
        let rowArray = dataRows[row]
        var command = assembleSQLITEInsertBase()
        command += "VALUES ("
        
        for (index, value) in rowArray.enumerated() {
            if value is NSNull {
                command += "NULL"
            } else {
                command += "\"\(value)\""
            }
            
            if index < rowArray.count - 1 {
                command += ","
            } else {
                command += ")"
            }
        }
        
        return command
    }
    
    @objc func writeToSQLITE(_ db: OpaquePointer?, withError error: NSErrorPointer) -> Bool {
        guard let db = db else { return false }
        
        let schemaSQL = sqliteTableSchema()
        var errorMsg: UnsafeMutablePointer<Int8>?
        
        if sqlite3_exec(db, schemaSQL, nil, nil, &errorMsg) != SQLITE_OK {
            if let error = error {
                error.pointee = NSError(domain: "PTFMXMLParser",
                                      code: 1,
                                      userInfo: [NSLocalizedDescriptionKey: String(cString: errorMsg!)])
            }
            sqlite3_free(errorMsg)
            return false
        }
        
        for row in 0..<dataRows.count {
            let command = sqliteCommand(forRow: row)
            if sqlite3_exec(db, command, nil, nil, &errorMsg) != SQLITE_OK {
                if let error = error {
                    error.pointee = NSError(domain: "PTFMXMLParser",
                                          code: 2,
                                          userInfo: [NSLocalizedDescriptionKey: String(cString: errorMsg!)])
                }
                sqlite3_free(errorMsg)
                return false
            }
        }
        
        return true
    }
    
    // MARK: - Private Methods
    
    private func assembleSQLITEInsertBase() -> String {
        var base = "INSERT INTO \"\(tableName)\" ("
        
        for (index, field) in tableDefinition.enumerated() {
            if let name = field["NAME"] {
                if index == tableDefinition.count - 1 {
                    base += "\(name)) "
                } else {
                    base += "\(name),"
                }
            }
        }
        
        return base
    }
}
